eknrenderer/eknr-renderer.c

```diff
@@ -0,0 +1,1306 @@
+
+#include "config.h"
+
+#include <string.h>
+#include <stdlib.h>
+
+#include <endless/endless.h>
+#include <glib/gi18n.h>
+#include <json-glib/json-glib.h>
+#include <mustache.h>
+
+#include "eknr-boxed-set-info.h"
+#include "eknr-errors.h"
+#include "eknr-renderer.h"
+
+/**
+ * SECTION:renderer
+ * @title: Renderer
+ * @short_description: A post-processing renderer for HTML article content
+ *
+ * The renderer is responsible for adding some final postprocessing to
+ * articles on the client side before it is displayed to the user. What
+ * postprocessing happens depends on the article's source.
+ */
+struct _EknrRenderer
+{
+  GObject parent_instance;
+};
+
+typedef struct _EknrRendererPrivate
+{
+  GHashTable *cache;
+} EknrRendererPrivate;
+
+G_DEFINE_TYPE_WITH_PRIVATE (EknrRenderer,
+                            eknr_renderer,
+                            G_TYPE_OBJECT)
+
+typedef struct _RendererMustacheData {
+  GVariantDict  *variables;
+  GError       **error; /* non-owned */
+
+  mustache_str_ctx input;
+  mustache_str_ctx output;
+
+  const char *section_variable;
+} RendererMustacheData;
+
+static RendererMustacheData *
+renderer_mustache_data_new (GVariantDict  *variables,
+                            GError       **error,
+                            const char    *input)
+{
+  RendererMustacheData *data = g_new0 (RendererMustacheData, 1);
+  data->variables = variables ? g_variant_dict_ref (variables) : NULL;
+  data->error = error;
+  data->input.string = input ? g_strdup (input) : NULL;
+  data->input.offset = 0;
+  data->output.string = NULL;
+  data->output.offset = 0;
+
+  return data;
+}
+
+static void
+renderer_mustache_data_free (RendererMustacheData *data)
+{
+  g_clear_pointer (&data->variables, g_variant_dict_unref);
+  g_clear_pointer (&data->output.string, free);
+  g_clear_pointer (&data->input.string, g_free);
+  g_free (data);
+}
+
+G_DEFINE_AUTOPTR_CLEANUP_FUNC (RendererMustacheData,
+                               renderer_mustache_data_free)
+
+static void
+free_mustache_template (mustache_template_t *template)
+{
+  mustache_api_t api = {
+    .freedata = NULL
+  };
+
+  mustache_free (&api, template);
+}
+
+G_DEFINE_AUTOPTR_CLEANUP_FUNC (mustache_template_t, free_mustache_template)
+
+static uintmax_t
+_renderer_read_from_closure (mustache_api_t *api,
+                             void           *userdata,
+                             char           *buffer,
+                             uintmax_t       buffer_size)
+{
+  RendererMustacheData *data = userdata;
+  return mustache_std_strread (api, (void *) &data->input, buffer, buffer_size);
+}
+
+static uintmax_t
+_renderer_write_to_closure (mustache_api_t *api,
+                            void           *userdata,
+                            const char     *buffer,
+                            uintmax_t       buffer_size)
+{
+  RendererMustacheData *data = userdata;
+  return mustache_std_strwrite (api, (void *) &data->output, (char *) buffer, buffer_size);
```

Comment:
This `(char *)` cast seems fishy to me. If the buffer is being written to, then it shouldn't be `const` when it's passed in to this function.

Reply:
The `(char *)` is fishy, I agree, but I had to do that because it is what mustache's API demands (I could fix that upstream but I don't want to break compatibility with other projects downstream that are using this library).

Reply:
Anyway, for what its worth, that buffer is not being written to. It is data->output that gets written to, but I agree the way the function is declared is misleading. I'll add a comment.

Reply:
OK, got it. A comment would be nice.